h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/matrix/RemoteMatrixLoader.java

```diff
@@ -0,0 +1,104 @@
+package hex.tree.xgboost.matrix;
+
+import hex.tree.xgboost.task.XGBoostUploadMatrixTask;
+import ml.dmlc.xgboost4j.java.util.BigDenseMatrix;
+import water.Key;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class RemoteMatrixLoader extends MatrixLoader {
+
+    static abstract class RemoteMatrix {
+        XGBoostUploadMatrixTask.MatrixData data;
+        
+        abstract MatrixLoader.DMatrixProvider make();
+    }
+    
+    static class RemoteSparseMatrix extends RemoteMatrix {
+        
+        final SparseMatrixDimensions dims;
+        final SparseMatrix matrix;
+
+        RemoteSparseMatrix(SparseMatrixDimensions dims) {
+            this.dims = dims;
+            this.matrix = SparseMatrixFactory.allocateCSRMatrix(dims);
+        }
+
+        @Override
+        MatrixLoader.DMatrixProvider make() {
+            return SparseMatrixFactory.toDMatrix(matrix, dims, data.actualRows, data.shape, data.resp, data.weights, data.offsets);
```

Comment:
is it somehow ensured that _data_ is defined when this method is called?

Reply:
yes, this should be called only after the matrix was completely uploaded and matrixData was called for this matrix